File: 0010-regular-expression-matching/0010-regular-expression-matching.py
"""
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        table = [[False] * (len(p)+1) for _ in range(len(s)+1)]
        table[0][0] = True
        
        for j in range(2, len(p)+1):
            if p[j-1] == "*":
                table[0][j] = table[0][j-2]
        
        for i in range(1, len(s)+1):
            for j in range(1, len(p)+1):
                if s[i-1] == p[j-1] or p[j-1] == ".":
                    table[i][j] = table[i-1][j-1]
                elif p[j-1] == "*":
                    empty = table[i][j-2]
                    nonempty = table[i-1][j] and (s[i-1] == p[j-2] or p[j-2] == ".")
                    table[i][j] = empty or nonempty
                    
        return table[-1][-1]
"""

class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        self.cache = {
            ("", ""): True,
        }
        return self.is_match(s, p)
    
    def is_match(self, s: str, p: str) -> bool:
        if self.cache.get((s, p)) != None:
            return self.cache[(s, p)]
        if p == "":
            return False
        
        match = False
        if p[-1] == "*":
            idx = len(s)-1
            match = match or self.is_match(s, p[:-2])
            while 0 <= idx and (s[idx] == p[-2] or p[-2] == "."):
                match = match or self.is_match(s[:idx], p[:-2])
                idx -= 1
        elif s != "" and (s[-1] == p[-1] or p[-1] == "."):
            match = match or self.is_match(s[:-1], p[:-1])
            
        self.cache[(s, p)] = match
        return match

File: 0010-regular-expression-matching/0010-regular-expression-matching.go
/*
func isMatch(s string, p string) bool {
    table := make([][]bool, len(s)+1)
    for i := 0; i < len(table); i++ {
        table[i] = make([]bool, len(p)+1)
    }
    table[0][0] = true
    
    for j := 2; j < len(p)+1; j++ {
        if p[j-1] == '*' {
            table[0][j] = table[0][j-2]
        }
    }
    
    for i := 1; i < len(s)+1; i++ {
        for j := 1; j < len(p)+1; j++ {
            if s[i-1] == p[j-1] || p[j-1] == '.' {
                table[i][j] = table[i-1][j-1]
            } else if p[j-1] == '*' {
                empty := table[i][j-2]
                nonempty := (s[i-1] == p[j-2] || p[j-2] == '.') && table[i-1][j]
                table[i][j] = empty || nonempty
            }
        }
    }
    
    return table[len(s)][len(p)]
}
*/

type pair struct {
    s, p string
}

func isMatch(s string, p string) bool {
    cache := map[pair]bool{
        pair{"", ""}: true,
    }
    
    var matches func(s, p string) bool
    matches = func(s, p string) bool {
        if cache[pair{s, p}] {
            return true
        }
        if p == "" {
            return false
        }
        
        match := false
        if p[len(p)-1] == '*' {
            match = match || matches(s, p[:len(p)-2])
            idx := len(s)-1
            for 0 <= idx && (s[idx] == p[len(p)-2] || p[len(p)-2] == '.') {
                match = match || matches(s[:idx], p[:len(p)-2])
                idx--
            }
        } else if s != "" && (s[len(s)-1] == p[len(p)-1] || p[len(p)-1] == '.') {
            match = match || matches(s[:len(s)-1], p[:len(p)-1])
        }
        
        cache[pair{s, p}] = match
        return match
    }
    
    return matches(s, p)
}